base/src/error.rs

```diff
@@ -45,72 +46,54 @@ impl<T: fmt::Display + fmt::Debug + Any> StdError for Errors<T> {
     }
 }
 
-
-#[derive(Debug)]
-struct SourceContext<E> {
-    context: String,
-    error: Spanned<E>,
-}
-
-fn extract_context<E>(lines: &[&str], error: Spanned<E>) -> SourceContext<E> {
-    SourceContext {
-        context: lines.get((error.span.start.line - 1) as usize)
-            .cloned()
-            .unwrap_or("N/A")
-            .to_string(),
-        error: error,
-    }
-}
-
 /// Error type which contains information of which file and where in the file the error occured
 #[derive(Debug)]
-pub struct InFile<E> {
-    file: String,
-    error: Errors<SourceContext<E>>,
+pub struct InFile {
+    message: String,
 }
 
-impl<E> InFile<E> {
+impl InFile {
     /// Creates a new `InFile` error which states that the error occured in `file` using the file
-    /// contents in `contents` to provide a context to the span.
-    pub fn new(file: String, contents: &str, error: Errors<Spanned<E>>) -> InFile<E> {
-        let lines: Vec<_> = contents.lines().collect();
-        InFile {
-            file: file,
-            error: Errors {
-                errors: error.errors
-                    .into_iter()
-                    .map(|error| extract_context(&lines, error))
-                    .collect(),
-            },
+    /// contents in `source` to provide a context to the span.
+    pub fn new<E: fmt::Display>(file: &str,
+                                source: &Source,
+                                error: Errors<Spanned<E, BytePos>>)
+                                -> InFile {
+        use std::fmt::Write;
+
+        let mut message = String::new();
+
+        for error in &error.errors {
+            let start = source.location(error.span.start).unwrap();
+            let end = source.location(error.span.end).unwrap();
+            let (_, line) = source.line(start.line).unwrap();
+
+            write!(message, "{}:{}\n{}\n", file, error.value, line).unwrap();
+
+            for _ in 1..start.column.to_usize() {
+                write!(message, " ").unwrap();
+            }
+
+            write!(message, "^").unwrap();
+
+            for _ in start.column.to_usize()..(end.column.to_usize() - 1) {
+                write!(message, "~").unwrap();
+            }
+
+            writeln!(message, "").unwrap();
```

Comment:
Any reason why you write all errors to a string straight away? I find it quite nice to keep errors typed as far as possible so it is possible to match on them and maybe have multiple ways of displaying them.